File: src/deadlock_build_sync/snapshot.py
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import StrEnum
from typing import Any


class MatchMode(StrEnum):
    """Supported public matchmaking populations."""

    RANKED = "ranked"
    UNRANKED = "unranked"

    @classmethod
    def parse(cls, value: str) -> MatchMode:
        try:
            return cls(value.strip().casefold())
        except ValueError as error:
            choices = ", ".join(mode.value for mode in cls)
            raise ValueError(
                f"unknown match mode {value!r}; choose {choices}"
            ) from error


class EvidenceUnit(StrEnum):
    """Units carried by exported observational evidence."""

    ASSET = "asset"
    PURCHASE_EVENT = "purchase_event"
    ELIGIBLE_APPEARANCE = "eligible_player_appearance"
    ABILITY_PATH = "ability_path"
    ABILITY_DECISION = "ability_decision_reached"
    HERO_APPEARANCE = "hero_appearance"
    HERO_ENEMY_PAIR = "hero_enemy_pair"
    GAME = "game"
    ITEM_FLOW_PAIR = "adjacent_phase_item_pair"


@dataclass(frozen=True)
class EpochBoundary:
    """Named start of one independently versioned evidence regime."""

    identity: str
    start_timestamp: int

    def __post_init__(self) -> None:
        """Validate identity and timestamp.

        Raises:
            ValueError: If the epoch identity or timestamp is invalid.

        """
        if not self.identity.strip():
            raise ValueError("epoch identity must not be empty")
        if self.start_timestamp < 0:
            raise ValueError("epoch start timestamp must be non-negative")

    def as_dict(self) -> dict[str, str | int]:
        return {
            "identity": self.identity,
            "start_timestamp": self.start_timestamp,
        }


@dataclass(frozen=True)
class EpochSet:
    """Independent boundaries that define one coherent evidence regime."""

    mechanics: EpochBoundary
    matchmaking: EpochBoundary
    map_objectives: EpochBoundary
    telemetry: EpochBoundary

    def as_dict(self) -> dict[str, dict[str, str | int]]:
        return {
            "mechanics": self.mechanics.as_dict(),
            "matchmaking": self.matchmaking.as_dict(),
            "map_objectives": self.map_objectives.as_dict(),
            "telemetry": self.telemetry.as_dict(),
        }

    @property
    def analysis_start_timestamp(self) -> int:
        return max(
            self.mechanics.start_timestamp,
            self.matchmaking.start_timestamp,
            self.map_objectives.start_timestamp,
            self.telemetry.start_timestamp,
        )


@dataclass(frozen=True)
class OutcomePolicy:
    """Outcome exclusions requested by the build-analysis contract."""

    exclude_not_scored: bool = True
    exclude_penalized: bool = True
    exclude_party_penalized: bool = True
    exclude_abandoned: bool = True
    exclude_unrewarded: bool = True
    exclude_low_priority: bool = True
    exclude_new_player: bool = True

    def as_dict(self, *, enforced: bool) -> dict[str, bool]:
        return {
            "exclude_not_scored": self.exclude_not_scored,
            "exclude_penalized": self.exclude_penalized,
            "exclude_party_penalized": self.exclude_party_penalized,
            "exclude_abandoned": self.exclude_abandoned,
            "exclude_unrewarded": self.exclude_unrewarded,
            "exclude_low_priority": self.exclude_low_priority,
            "exclude_new_player": self.exclude_new_player,
            "enforced_by_source": enforced,
        }


def canonical_json(value: Any) -> bytes:
    """Encode a value for stable evidence fingerprints.

    Returns:
        Stable UTF-8 JSON bytes.

    """
    return json.dumps(
        value,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode()


def sha256_json(value: Any) -> str:
    return hashlib.sha256(canonical_json(value)).hexdigest()


@dataclass(frozen=True)
class EvidenceRecord:
    """One immutable API response and its declared analytic semantics."""

    path: str
    parameters: dict[str, Any]
    fetched_at: str
    sha256: str
    byte_count: int
    unit: EvidenceUnit
    backend_grain: str
    fallback_behavior: str
    warnings: tuple[str, ...] = ()

    def as_dict(self) -> dict[str, Any]:
        return {
            "path": self.path,
            "parameters": self.parameters,
            "fetched_at": self.fetched_at,
            "sha256": self.sha256,
            "byte_count": self.byte_count,
            "unit": self.unit.value,
            "backend_grain": self.backend_grain,
            "fallback_behavior": self.fallback_behavior,
            "warnings": list(self.warnings),
        }

    def identity_dict(self) -> dict[str, Any]:
        """Return stable source identity without wall-clock fetch metadata.

        Returns:
            The request, response, and semantics fields that define compatibility.

        """
        payload = self.as_dict()
        payload.pop("fetched_at")
        return payload


@dataclass
class EvidenceRecorder:
    """Collect exact request/response identities for one generation run."""

    records: list[EvidenceRecord] = field(default_factory=list)
    _semantics: dict[str, tuple[EvidenceUnit, str, str, tuple[str, ...]]] = field(
        default_factory=dict,
        repr=False,
    )

    def declare(
        self,
        path: str,
        *,
        unit: EvidenceUnit,
        backend_grain: str,
        fallback_behavior: str = "none",
        warnings: tuple[str, ...] = (),
    ) -> None:
        if not backend_grain.strip():
            raise ValueError("backend grain must not be empty")
        if not fallback_behavior.strip():
            raise ValueError("fallback behavior must not be empty")
        self._semantics[path] = (
            unit,
            backend_grain,
            fallback_behavior,
            warnings,
        )

    def record(
        self,
        path: str,
        parameters: dict[str, Any],
        raw: bytes,
        *,
        fetched_at: datetime | None = None,
    ) -> EvidenceRecord:
        semantics = self._semantics.get(path)
        if semantics is None:
            raise ValueError(f"evidence semantics were not declared for {path}")
        record = EvidenceRecord(
            path=path,
            parameters=dict(sorted(parameters.items())),
            fetched_at=(fetched_at or datetime.now(UTC)).isoformat(),
            sha256=hashlib.sha256(raw).hexdigest(),
            byte_count=len(raw),
            unit=semantics[0],
            backend_grain=semantics[1],
            fallback_behavior=semantics[2],
            warnings=semantics[3],
        )
        self.records.append(record)
        return record


@dataclass(frozen=True)
class SnapshotManifest:
    """Frozen source, cohort, and request identities for a generated policy."""

    client_version: int
    as_of_timestamp: int
    created_at: str
    match_mode: MatchMode
    game_mode: str
    rank_range: dict[str, object]
    rank_labels_sha256: str
    patch: dict[str, Any]
    epochs: EpochSet
    outcome_policy: OutcomePolicy
    outcome_policy_enforced: bool
    records: tuple[EvidenceRecord, ...]
    build_tags_sha256: str = ""
    warnings: tuple[str, ...] = ()

    def __post_init__(self) -> None:
        """Validate cross-field snapshot invariants.

        Raises:
            ValueError: If the manifest is incomplete or temporally incoherent.

        """
        if self.client_version <= 0:
            raise ValueError("client version must be positive")
        if self.game_mode != "normal":
            raise ValueError("only the normal Deadlock ruleset is supported")
        if self.as_of_timestamp < self.epochs.analysis_start_timestamp:
            raise ValueError("as-of cutoff precedes a required epoch boundary")
        if not self.records:
            raise ValueError("snapshot manifest must contain source records")
        if len(self.build_tags_sha256) != 64:
            raise ValueError("snapshot manifest has no valid build-tag fingerprint")

    def _payload(self, *, identity: bool = False) -> dict[str, Any]:
        return {
            "schema_version": 1,
            "client_version": self.client_version,
            "as_of_timestamp": self.as_of_timestamp,
            **({} if identity else {"created_at": self.created_at}),
            "match_mode": self.match_mode.value,
            "game_mode": self.game_mode,
            "rank_range": self.rank_range,
            "rank_labels_sha256": self.rank_labels_sha256,
            "build_tags_sha256": self.build_tags_sha256,
            "patch": self.patch,
            "epochs": self.epochs.as_dict(),
            "outcome_policy": self.outcome_policy.as_dict(
                enforced=self.outcome_policy_enforced
            ),
            "records": [
                record.identity_dict() if identity else record.as_dict()
                for record in self.records
            ],
            "warnings": list(self.warnings),
        }

    @property
    def snapshot_id(self) -> str:
        return sha256_json(self._payload(identity=True))

    def as_dict(self) -> dict[str, Any]:
        payload = self._payload()
        payload["snapshot_id"] = self.snapshot_id
        return payload

File: crates/deadlock-data/src/snapshot.rs
use chrono::DateTime;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::error::{Error, Result};
use crate::evidence::{EpochSet, EvidenceRecord, MatchMode, validate_sha256};
use crate::json::fingerprint;

const OUTCOME_EXCLUSIONS: [&str; 7] = [
    "exclude_not_scored",
    "exclude_penalized",
    "exclude_party_penalized",
    "exclude_abandoned",
    "exclude_unrewarded",
    "exclude_low_priority",
    "exclude_new_player",
];

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SnapshotContent {
    pub schema_version: u32,
    pub client_version: u64,
    pub as_of_timestamp: i64,
    pub created_at: String,
    pub match_mode: MatchMode,
    pub game_mode: String,
    pub rank_range: Map<String, Value>,
    pub rank_labels_sha256: String,
    pub build_tags_sha256: String,
    pub patch: Map<String, Value>,
    pub epochs: EpochSet,
    pub outcome_policy: Map<String, Value>,
    pub records: Vec<EvidenceRecord>,
    pub warnings: Vec<String>,
}

#[derive(Clone, Debug)]
pub struct SnapshotManifest {
    content: SnapshotContent,
    identifier: String,
}

impl SnapshotManifest {
    /// # Errors
    /// Returns an error for incomplete source identities or inconsistent evidence timestamps.
    pub fn new(content: SnapshotContent) -> Result<Self> {
        validate_content(&content)?;
        let mut identity = serde_json::to_value(&content)?;
        remove_fetch_times(&mut identity)?;
        let identifier = fingerprint(&identity)?;
        Ok(Self {
            content,
            identifier,
        })
    }

    /// # Errors
    /// Returns an error for an invalid manifest or a mismatched snapshot fingerprint.
    pub fn from_document(document: &Value) -> Result<Self> {
        let identifier = document["snapshot_id"]
            .as_str()
            .ok_or_else(|| Error::new("Snapshot manifest has no snapshot ID"))?;
        let mut payload = document.clone();
        payload
            .as_object_mut()
            .ok_or_else(|| Error::new("Snapshot manifest must be an object"))?
            .remove("snapshot_id");
        let content = serde_json::from_value(payload)?;
        let manifest = Self::new(content)?;
        if identifier != manifest.identifier {
            return Err(Error::new(
                "Snapshot ID does not match the source identities",
            ));
        }
        Ok(manifest)
    }

    #[must_use]
    pub const fn content(&self) -> &SnapshotContent {
        &self.content
    }

    #[must_use]
    pub fn identifier(&self) -> &str {
        &self.identifier
    }

    /// # Errors
    /// Returns an error when epoch serialization fails.
    pub fn cohort_record(&self) -> Result<Map<String, Value>> {
        Ok(Map::from_iter([
            ("match_mode".into(), self.content.match_mode.as_str().into()),
            ("game_mode".into(), self.content.game_mode.clone().into()),
            ("rank_range".into(), self.content.rank_range.clone().into()),
            (
                "as_of_timestamp".into(),
                self.content.as_of_timestamp.into(),
            ),
            ("epochs".into(), serde_json::to_value(&self.content.epochs)?),
        ]))
    }

    /// # Errors
    /// Returns an error when a rank bound has no display label.
    pub fn rank_identity(&self) -> Result<String> {
        let format_rank = |name: &str| -> Result<String> {
            let rank = self
                .content
                .rank_range
                .get(name)
                .ok_or_else(|| Error::new("Snapshot rank bound is missing"))?;
            let label = rank["label"]
                .as_str()
                .ok_or_else(|| Error::new("Snapshot rank label is missing"))?;
            let badge = rank["badge_id"]
                .as_u64()
                .ok_or_else(|| Error::new("Snapshot rank badge is missing"))?;
            Ok(format!("{label} [{badge}]"))
        };
        let minimum = format_rank("minimum")?;
        let maximum = format_rank("maximum")?;
        Ok(if minimum == maximum {
            minimum
        } else {
            format!("{minimum}–{maximum}")
        })
    }

    /// # Errors
    /// Returns an error when manifest serialization fails.
    pub fn to_document(&self) -> Result<Value> {
        let mut document = serde_json::to_value(&self.content)?;
        document["snapshot_id"] = self.identifier.clone().into();
        Ok(document)
    }
}

fn validate_content(content: &SnapshotContent) -> Result<()> {
    if content.schema_version != 1 || content.client_version == 0 || content.game_mode != "normal" {
        return Err(Error::new(
            "Snapshot requires schema 1, a positive client version, and the normal game mode",
        ));
    }
    content.epochs.validate()?;
    if content.as_of_timestamp < content.epochs.analysis_start_timestamp() {
        return Err(Error::new(
            "Snapshot cutoff precedes a required epoch boundary",
        ));
    }
    if content.records.is_empty() {
        return Err(Error::new("Snapshot manifest has no source records"));
    }
    validate_sha256(&content.rank_labels_sha256, "Rank catalog")?;
    validate_sha256(&content.build_tags_sha256, "Build tag catalog")?;
    validate_rank_range(&content.rank_range)?;
    validate_outcome_policy(&content.outcome_policy)?;
    DateTime::parse_from_rfc3339(&content.created_at)
        .map_err(|error| Error::new(format!("Snapshot creation timestamp is invalid: {error}")))?;
    for record in &content.records {
        record.validate()?;
    }
    Ok(())
}

fn validate_rank_range(value: &Map<String, Value>) -> Result<()> {
    let rank = |name: &str| -> Result<crate::rank::Rank> {
        let badge = value
            .get(name)
            .and_then(|value| value.get("badge_id"))
            .and_then(Value::as_u64)
            .ok_or_else(|| Error::new(format!("Snapshot rank range has no {name} badge")))?;
        crate::rank::Rank::try_from(u16::try_from(badge)?)
    };
    crate::rank::RankRange {
        minimum: rank("minimum")?,
        maximum: rank("maximum")?,
    }
    .validate()?;
    Ok(())
}

fn validate_outcome_policy(value: &Map<String, Value>) -> Result<()> {
    for name in OUTCOME_EXCLUSIONS.into_iter().chain(["enforced_by_source"]) {
        if value.get(name).and_then(Value::as_bool).is_none() {
            return Err(Error::new(format!(
                "Snapshot outcome policy has no boolean {name} field"
            )));
        }
    }
    Ok(())
}

fn remove_fetch_times(document: &mut Value) -> Result<()> {
    let object = document
        .as_object_mut()
        .ok_or_else(|| Error::new("Snapshot identity is not an object"))?;
    object.remove("created_at");
    let records = object
        .get_mut("records")
        .and_then(Value::as_array_mut)
        .ok_or_else(|| Error::new("Snapshot identity has no source record list"))?;
    for record in records {
        record
            .as_object_mut()
            .ok_or_else(|| Error::new("Snapshot source record is not an object"))?
            .remove("fetched_at");
    }
    Ok(())
}

#[must_use]
pub fn default_outcome_policy(enforced: bool) -> Map<String, Value> {
    let mut policy = OUTCOME_EXCLUSIONS
        .into_iter()
        .map(|name| (name.to_owned(), Value::Bool(true)))
        .collect::<Map<_, _>>();
    policy.insert("enforced_by_source".into(), enforced.into());
    policy
}
